From 364695a4a3eb3b2416c82c1fcb88c33081c20fae Mon Sep 17 00:00:00 2001
From: Venkateshwar Rao Gannavarapu <venkateshwar.rao.gannavarapu@xilinx.com>
Date: Mon, 2 Nov 2020 01:01:48 -0800
Subject: [PATCH 1737/1852] v4l: xilinx: sdirxss: Fix PLL reference clock to
 148.5MHz for SD mode

commit 4f8b566cf1188d40ba08812f6ea9ac212fab3f9b from
https://github.com/Xilinx/linux-xlnx.git

This patch sets the PLL reference clock to 148.5MHZ for SD mode, as
mentioned in SDI-RX subsystem product guide PG290, clocking section.

Without this patch 720x486i@59.94 captures green screen for picxo design

Signed-off-by: Venkateshwar Rao Gannavarapu <venkateshwar.rao.gannavarapu@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-sdirxss.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-sdirxss.c b/drivers/media/platform/xilinx/xilinx-sdirxss.c
index 793a1b44859a..f1afc561b4dd 100644
--- a/drivers/media/platform/xilinx/xilinx-sdirxss.c
+++ b/drivers/media/platform/xilinx/xilinx-sdirxss.c
@@ -1038,8 +1038,14 @@ static void xsdirxss_set_gtclk(struct xsdirxss_state *state)
 	gtclk = core->clks[1].clk;
 	is_frac = state->frame_interval.numerator == 1001 ? 1 : 0;
 
-	/* calcualte clkrate */
-	if (!is_frac)
+	/*
+	 * PLL ref clock is 148.5MHz for integer frame rates
+	 * and 148.35MHz for fractional frame rates.
+	 * For SD mode its always 148.5MHz for integer & fractional.
+	 * Please refer to Table 5-2 in PG290
+	 * https://www.xilinx.com/support/documentation/ip_documentation/v_smpte_uhdsdi_rx_ss/v2_0/pg290-v-smpte-uhdsdi-rx-ss.pdf
+	 */
+	if (!is_frac || mode == XSDIRX_MODE_SD_MASK)
 		clkrate = CLK_INT;
 	else
 		clkrate = (CLK_INT * 1000) / 1001;
-- 
2.31.1

